Port improved pt-pmp

svetasmirnova opened the pull request:
Initial port of https://github.com/Percona-Lab/pt-pmp:
- Added bin/pt-eustack-resolver
- Ported changes to bin/pt-pmp

- [ ] The contributed code is licensed under GPL v2.0
- [ ] Contributor Licence Agreement (CLA) is signed
- [ ] util/update-modules has been ran
- [ ] Documentation updated
- [ ] Test suite update

svetasmirnova committed:
Port improved pt-pmp

Initial port of https://github.com/Percona-Lab/pt-pmp:
- Added bin/pt-eustack-resolver
- Ported changes to bin/pt-pmp

svetasmirnova committed:
Port improved pt-pmp

- Updated documentation for pt-eustack-resolver and pt-pmp
- Added pt-eustack-resolver, pt-pmp and other missed tools into docs/percona-toolkit.pod
- Removed quickstack support from pt-pmp. We do not need to port this
  improvement, because quickstack is not actively developed and does not
  create useful stack traces for Percona Server for MySQL 8.0.

svetasmirnova committed:
Port improved pt-pmp

- Added test case for option --tids

svetasmirnova committed:
Port improved pt-pmp

- Made --readnever for gdb optional
- Re-arranged code, related to dumper selection

svetasmirnova committed:
Port improved pt-pmp

- Added tests for alternative dumpers

svetasmirnova committed:
Port improved pt-pmp

- Added test cases for proper reading of eu-stack dumps (when option
  --dumper not default)

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,253 @@
+
+  if ($pid =~ /^[0-9]+$/) {
+    open F, '<', "/proc/$pid/maps"
+        or die "Failed to open /proc/$pid/maps: $!\n";
```
according to perl best practices it is always better to re-define filehandler for opening file
open my  $FH, '<' "<filename>"

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,253 @@
+    open F, '<', "/proc/$pid/maps"
+        or die "Failed to open /proc/$pid/maps: $!\n";
+  } else {
+    open F, '<', $pid
```
the same as previous comment

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,253 @@
+
+my $syms= { };
+
+sub get_image {
+  my ($addr)= @_;
+  my $e;
+  for $e (@$mmap) {
+    next if $e->{E} <= $addr;
+    last if $e->{S} > $addr;
+    # Found, look up.  
+      return $e->{F};
+  }
+  return "";
+}
+
```
sub get_image {
    my ($addr) = @_;

    # Ensure addr is defined
    die "Address is undefined" unless defined $addr;

    # Check if the global memory map is defined and is an array reference
    die "Global memory map is undefined or not an array reference"
        unless defined $memory_map and ref($memory_map) eq 'ARRAY';

    foreach my $entry (@$memory_map) {
        next if $entry->{E} <= $addr;
        last if $entry->{S} > $addr;
        # Found, return file
        return $entry->{F};
    }

    return "";
}

I would suggest trying something like this

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,253 @@
+#  printf "0x%x - 0x%x (0x%x): %s\n", $_->{S}, $_->{E}, $_->{B}, $_->{F};
+#}
+
+open (STACK_TRACE, "eu-stack -q -p $pid 2>/dev/null|") or die "open(): $!";
+my @lines= <STACK_TRACE>;
+close(STACK_TRACE);
```
```
open(my $stack_trace, '-|', "eu-stack -q -p $pid 2>/dev/null") or die "open() failed: $!";
my @lines = <$stack_trace>;
close($stack_trace) or die "close() failed: $!";
```

svetasmirnova committed:
Port improved pt-pmp

- Suggestions by Evgeniy

svetasmirnova commented:
@EvgeniyPatlan implemented, please re-check

ihanick commented:
I've tried --dumper eu against PS 8.0 and pt-pmp shows exactly the same or even logger time compare to gdb.
This is confirmed by the article https://www.percona.com/blog/evaluation-of-profiling-tools-for-pmp/
At the same time -q option reduces the dump file from 1.6 seconds to 0.3 seconds on the idle system.
bin/pt-eustack-resolver does the job, but when it's not clear why dumper eu mode exists

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,258 @@
+   my ($pid)= @_;
+
+   if ($pid =~ /^[0-9]+$/) {
+      open $FH, '<', "/proc/$pid/maps"
+         or die "Failed to open /proc/$pid/maps: $!\n";
```
open my $FH

svetasmirnova replied:
This would not work, because this code is in the scope of the if statement.

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,258 @@
+      open $FH, '<', "/proc/$pid/maps"
+         or die "Failed to open /proc/$pid/maps: $!\n";
+   } else {
+      open $FH, '<', $pid
+         or die "Failed to open saved map file '$pid': $!\n";
```
open my $FH

svetasmirnova replied:
This would not work, because this code is in the scope of the if statement.

EvgeniyPatlan commented on bin/pt-eustack-resolver:
```diff
@@ -0,0 +1,258 @@
+   die "Global memory map is undefined or not an array reference"
+      unless defined $mmap and ref($mmap) eq 'ARRAY';
+
+   my $e;
+   for $e (@$mmap) {
```
for my $e (@$mmap) {

svetasmirnova replied:
Done

svetasmirnova committed:
Port improved pt-pmp

- Variable scope fixes

ihanick commented:
> I've tried --dumper eu against PS 8.0 and pt-pmp shows exactly the same or even logger time compare to gdb. This is confirmed by the article https://www.percona.com/blog/evaluation-of-profiling-tools-for-pmp/ At the same time -q option reduces the dump file from 1.6 seconds to 0.3 seconds on the idle system. bin/pt-eustack-resolver does the job, but when it's not clear why dumper eu mode exists

On the overloaded system with 80 connections from sysbench:
time bin/pt-pmp --dumper eu
real	0m10.343s

+ sysbench --db-driver=mysql --table-size=100000 --tables=8 --threads=80 --mysql-host=172.20.0.2 --mysql-user=root '--mysql-password=verysecretpassword1^' --mysql-db=sbtest --time=60 --report-interval=2 /usr/share/sysbench/oltp_read_write.lua run
sysbench 1.0.20 (using bundled LuaJIT 2.1.0-beta2)
[ 2s ] thds: 80 tps: 302.93 qps: 6799.11 (r/w/o: 4793.57/1360.22/645.33) lat (ms,95%): 350.33 err/s: 0.00 reconn/s: 0.00
[ 4s ] thds: 80 tps: 273.84 qps: 5328.09 (r/w/o: 3833.74/946.67/547.68) lat (ms,95%): 411.96 err/s: 0.00 reconn/s: 0.00
[ 6s ] thds: 80 tps: 193.12 qps: 4019.53 (r/w/o: 2703.70/929.59/386.24) lat (ms,95%): 1235.62 err/s: 0.00 reconn/s: 0.00
[ 8s ] thds: 80 tps: 197.03 qps: 3940.14 (r/w/o: 2758.44/787.63/394.06) lat (ms,95%): 1089.30 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 80 tps: 148.77 qps: 2972.32 (r/w/o: 2082.72/592.07/297.53) lat (ms,95%): 634.66 err/s: 0.00 reconn/s: 0.00
[ 12s ] thds: 80 tps: 162.99 qps: 3251.84 (r/w/o: 2281.88/643.97/325.98) lat (ms,95%): 1013.60 err/s: 0.00 reconn/s: 0.00
[ 14s ] thds: 80 tps: 124.00 qps: 2488.05 (r/w/o: 1736.03/504.01/248.00) lat (ms,95%): 1069.86 err/s: 0.00 reconn/s: 0.00
[ 16s ] thds: 80 tps: 112.67 qps: 2121.76 (r/w/o: 1577.42/318.99/225.35) lat (ms,95%): 1648.20 err/s: 0.00 reconn/s: 0.00
-> runs here [ 18s ] thds: 80 tps: 20.00 qps: 382.50 (r/w/o: 280.00/62.50/40.00) lat (ms,95%): 3095.38 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 80 tps: 105.43 qps: 2249.98 (r/w/o: 1476.01/563.12/210.86) lat (ms,95%): 3326.55 err/s: 0.00 reconn/s: 0.00

GDB:
time bin/pt-pmp --dumper gdb
Thu Jan 25 20:23:47 UTC 2024
bin/pt-pmp: line 752:  7633 Segmentation fault      (core dumped) gdb -ex 'set pagination 0' -ex 'thread apply all bt' -batch -p 5209
      6 nanosleep(libpthread.so.0),std::this_thread::sleep_for<long,std::ratio<1l,1000000l>>(this_thread_sleep.h:80),ut::wait_for<log_free_check_wait(log_t&)::<lambda(bool)>(ut0ut.ic:174),log_free_check_wait(log0chkp.cc:1267),log_free_check(log0chkp.h:273),log_free_check(log0chkp.h:259),log_free_check(log0chkp.h:312),row_upd(row0upd.cc:3184),row_upd_step(row0upd.cc:3317),row_update_for_mysql_using_upd_graph(row0mysql.cc:2857),ha_innobase::update_row(ha_innodb.cc:10379),handler::ha_update_row(handler.cc:8365),Sql_cmd_update::update_single_table(sql_update.cc:1002),Sql_cmd_dml::execute(sql_select.cc:793),mysql_execute_command(sql_parse.cc:4978),Prepared_statement::execute(sql_prepare.cc:3653),Prepared_statement::execute_loop(sql_prepare.cc:3062),mysqld_stmt_execute(sql_prepare.cc:1918),dispatch_command(sql_parse.cc:2035),do_command(sql_parse.cc:1501),handle_connection(connection_handler_per_thread.cc:308),pfs_spawn_thread(pfs.cc:3043),start_thread(libpthread.so.0),clone(libc.so.6)
      1 nanosleep(libpthread.so.0)

real	0m20.786s

+ sysbench --db-driver=mysql --table-size=100000 --tables=8 --threads=80 --mysql-host=172.20.0.2 --mysql-user=root '--mysql-password=verysecretpassword1^' --mysql-db=sbtest --time=60 --report-interval=2 /usr/share/sysbench/oltp_read_write.lua run
sysbench 1.0.20 (using bundled LuaJIT 2.1.0-beta2)

Running the test with following options:
Number of threads: 80
Report intermediate results every 2 second(s)
Initializing random number generator from current time


Initializing worker threads...

Threads started!

[ 2s ] thds: 80 tps: 1406.66 qps: 28783.67 (r/w/o: 20195.42/5736.86/2851.38) lat (ms,95%): 104.84 err/s: 0.00 reconn/s: 0.00
[ 4s ] thds: 80 tps: 1763.36 qps: 35234.29 (r/w/o: 24675.11/7031.96/3527.23) lat (ms,95%): 74.46 err/s: 0.00 reconn/s: 0.00
[ 6s ] thds: 80 tps: 1310.24 qps: 26315.36 (r/w/o: 18407.40/5286.48/2621.48) lat (ms,95%): 139.85 err/s: 0.00 reconn/s: 0.00
[ 8s ] thds: 80 tps: 1479.74 qps: 29534.74 (r/w/o: 20679.32/5896.45/2958.97) lat (ms,95%): 102.97 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 80 tps: 1257.98 qps: 25191.54 (r/w/o: 17631.68/5043.91/2515.95) lat (ms,95%): 132.49 err/s: 0.00 reconn/s: 0.00
[ 12s ] thds: 80 tps: 577.15 qps: 11440.91 (r/w/o: 8097.06/2189.06/1154.79) lat (ms,95%): 101.13 err/s: 0.00 reconn/s: 0.00
[ 14s ] thds: 80 tps: 120.30 qps: 2403.59 (r/w/o: 1684.26/478.72/240.61) lat (ms,95%): 1453.01 err/s: 0.00 reconn/s: 0.00
[ 16s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 18s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 22s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 24s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 26s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 28s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 30s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 32s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 34s ] thds: 80 tps: 0.50 qps: 154.75 (r/w/o: 7.01/146.73/1.00) lat (ms,95%): 21255.35 err/s: 0.00 reconn/s: 0.00
[ 36s ] thds: 80 tps: 652.84 qps: 12961.79 (r/w/o: 9087.25/2569.36/1305.18) lat (ms,95%): 21641.55 err/s: 0.00 reconn/s: 0.00
[ 38s ] thds: 80 tps: 753.19 qps: 15014.79 (r/w/o: 10597.18/2910.74/1506.88) lat (ms,95%): 846.57 err/s: 0.00 reconn/s: 0.00
[ 40s ] thds: 80 tps: 542.94 qps: 10910.26 (r/w/o: 7571.14/2253.74/1085.38) lat (ms,95%): 1050.76 err/s: 0.00 reconn/s: 0.00
[ 42s ] thds: 80 tps: 947.59 qps: 18899.73 (r/w/o: 13296.22/3707.84/1895.67) lat (ms,95%): 223.34 err/s: 0.00 reconn/s: 0.00
[ 44s ] thds: 80 tps: 786.39 qps: 15861.23 (r/w/o: 11003.42/3285.03/1572.77) lat (ms,95%): 1032.01 err/s: 0.00 reconn/s: 0.00
[ 46s ] thds: 80 tps: 1513.09 qps: 30280.37 (r/w/o: 21189.31/6064.87/3026.19) lat (ms,95%): 86.00 err/s: 0.00 reconn/s: 0.00


pt-eu-stack-resolver:
time bin/pt-eustack-resolver $(pgrep -xn mysqld)
Thread 5209 (LWP 5209):
#0 0x00007f60548cff41 in __poll from /usr/lib64/libc-2.28.so (deleted)
#1 0x0000000001212818 in Mysqld_socket_listener::listen_for_connection_event() at socket_connection.cc:1356:3
#2 0x0000000000f49258 in Connection_acceptor<Mysqld_socket_listener>::connection_event_loop() at connection_acceptor.h:64:75
#3 0x00007f60547e6d85 in __libc_start_main from /usr/lib64/libc-2.28.so (deleted)
#4 0x0000000000d776ce in _start from /usr/sbin/mysqld

Thread 5212 (LWP 5212):
#0 0x00007f60547e59bd in syscall from /usr/lib64/libc-2.28.so (deleted)
#1 0x00007f60558c4bbe in __io_getevents_0_4 from /usr/lib64/libaio.so.1.0.1
#2 0x00000000022ddd3b in LinuxAIOHandler::collect() at os0file.cc:2387:28
#3 0x00000000022de1b2 in LinuxAIOHandler::poll(fil_node_t**, void**, IORequest*) at os0file.cc:2535:3
#4 0x00000000022e59f4 in os_aio_linux_handler at os0file.cc:2583:3
#5 0x000000000255cb4d in fil_aio_wait(unsigned long) at fil0fil.cc:7981:3
#6 0x00000000022d86b0 in std::atomic<srv_shutdown_t>::load(std::memory_order) const at atomic:285:15
#7 0x00000000022d663e in std::__shared_ptr<std::__future_base::_State_baseV2, (__gnu_cxx::_Lock_policy)2>::operator bool() const at shared_ptr_base.h:1670:16
#8 0x00007f60551d6b13 in execute_native_thread_routine from /usr/lib64/libstdc++.so.6.0.25
#9 0x00007f60564641ca in ?? from /usr/lib64/libpthread-2.28.so (deleted)
#10 0x00007f60547e5e73 in __clone from /usr/lib64/libc-2.28.so (deleted)
....
real	0m1.617s
user	0m1.429s

[ 2s ] thds: 80 tps: 1320.08 qps: 27063.31 (r/w/o: 18979.39/5404.64/2679.29) lat (ms,95%): 134.90 err/s: 0.00 reconn/s: 0.00
[ 4s ] thds: 80 tps: 1514.34 qps: 30322.78 (r/w/o: 21222.74/6071.86/3028.18) lat (ms,95%): 101.13 err/s: 0.00 reconn/s: 0.00
[ 6s ] thds: 80 tps: 1329.38 qps: 26639.56 (r/w/o: 18646.29/5333.50/2659.76) lat (ms,95%): 116.80 err/s: 0.00 reconn/s: 0.00
[ 8s ] thds: 80 tps: 1245.96 qps: 24855.05 (r/w/o: 17419.37/4943.77/2491.91) lat (ms,95%): 139.85 err/s: 0.00 reconn/s: 0.00
[ 10s ] thds: 80 tps: 1610.44 qps: 32169.72 (r/w/o: 22522.10/6428.74/3218.87) lat (ms,95%): 87.56 err/s: 0.00 reconn/s: 0.00
[ 12s ] thds: 80 tps: 1290.03 qps: 25858.51 (r/w/o: 18088.86/5188.10/2581.55) lat (ms,95%): 102.97 err/s: 0.00 reconn/s: 0.00
[ 14s ] thds: 80 tps: 1185.53 qps: 23609.58 (r/w/o: 16530.90/4708.61/2370.06) lat (ms,95%): 108.68 err/s: 0.00 reconn/s: 0.00
[ 16s ] thds: 80 tps: 662.63 qps: 13247.13 (r/w/o: 9362.86/2557.51/1326.76) lat (ms,95%): 960.30 err/s: 0.00 reconn/s: 0.00
[ 18s ] thds: 80 tps: 547.47 qps: 10948.47 (r/w/o: 7664.63/2188.89/1094.95) lat (ms,95%): 831.46 err/s: 0.00 reconn/s: 0.00
[ 20s ] thds: 80 tps: 651.50 qps: 13040.00 (r/w/o: 9121.00/2616.00/1303.00) lat (ms,95%): 549.52 err/s: 0.00 reconn/s: 0.00
[ 22s ] thds: 80 tps: 1290.27 qps: 25789.90 (r/w/o: 17958.80/5252.06/2579.04) lat (ms,95%): 215.44 err/s: 0.00 reconn/s: 0.00
[ 24s ] thds: 80 tps: 828.18 qps: 16693.08 (r/w/o: 11687.50/3347.72/1657.86) lat (ms,95%): 303.33 err/s: 0.00 reconn/s: 0.00
[ 26s ] thds: 80 tps: 606.36 qps: 11959.83 (r/w/o: 8368.63/2380.97/1210.23) lat (ms,95%): 634.66 err/s: 0.00 reconn/s: 0.00
[ 28s ] thds: 80 tps: 754.67 qps: 15287.43 (r/w/o: 10697.40/3078.19/1511.84) lat (ms,95%): 434.83 err/s: 0.00 reconn/s: 0.00


time bin/pt-pmp --dumper gdb --readnever
real	0m4.088s
[ 10s ] thds: 80 tps: 331.55 qps: 6632.00 (r/w/o: 4642.70/1326.20/663.10) lat (ms,95%): 344.08 err/s: 0.00 reconn/s: 0.00
[ 12s ] thds: 80 tps: 0.00 qps: 0.00 (r/w/o: 0.00/0.00/0.00) lat (ms,95%): 0.00 err/s: 0.00 reconn/s: 0.00
[ 14s ] thds: 80 tps: 230.86 qps: 4615.26 (r/w/o: 3232.08/921.45/461.73) lat (ms,95%): 3773.42 err/s: 0.00 reconn/s: 0.00

With 500 threads, normal pt-pmp never finishes, with elfutils finsihes in 60 seconds without freezing connections
pt-eustack-resolver returns in 5 seconds under same conditions.

Conclusion:
Load makes elfutils useful, it's even better that queries are still processed during the stack trace.
GDB freezes the process at least for few seconds even with readnever / non-stop on / observer on.

ihanick commented on bin/pt-pmp:
```diff
@@ -671,22 +708,57 @@ main() {
+      	 if [ `_which eu-stack` ] && [ `_which pt-eustack-resolver` ]; then
+      	    DUMPER="pt-eustack-resolver ${OPT_PID}"
+      	 else
+      	 	die "Can't find eu-stack binary from elfutils or pt-eustack-resolver. Exiting"
```
Message is confusing. It's better to generate separate messages for eu-stack and for pt-eustack-resolver.
E.g. 
die "Can't find eu-stack binary from elfutils package"
and
die "Can't find pt-eustack-resolver tool in PATH."

If the percona toolkit is installed from tar.gz, pt-eustack-resolver could be not in $PATH, but in the same directory as pt-pmp

svetasmirnova replied:
Updated, please check.

ihanick reviewed (changes requested):
Please fix the error message if pt-eustack-resolver is not in the path and LGTM

svetasmirnova committed:
Port improved pt-pmp

- Updated error messages when eu-stack and/or pt-eustack-resolver tools
  are not found

ihanick reviewed (approved):
LGTM
